[homuntan02] iP

homuntan02 opened the pull request:
```ruby
 ____        _        
|  _ \ _   _| | _____ 
| | | | | | | |/ / _ \
| |_| | |_| |   <  __/
|____/ \__,_|_|\_\___|

```
Duke helps you keep track of things you need to do in 1 simple list. it has:

- **Straightforward** and **simple** commands
- Marker system to track which task is completed

Features include:

- [x] Managing Task
- [x] 3 different task types :
        -  To-Dos
        -  Deadlines
        -  Events

- [x] Searching for task with keywords
- [ ] Task Reminder (coming soon)

Here is the `main` method if you code in Java.
```ruby
public class Duke {
    public static void main(String[] args) {
        new Duke("./data/").run();
    }
}
```

Commit:
Add Gradle support

homuntan02 committed:
IP Level-1

homuntan02 committed:
IP Level-2

homuntan02 committed:
IP Level-3

homuntan02 committed:
IP Level-4 added ToDo, Events and Deadline

homuntan02 committed:
IP Level-5 Handling error

homuntan02 committed:
IP  Automated Text UI Testing

homuntan02 committed:
IP Leve-6 Delete

homuntan02 committed:
Level-7 save completed

homuntan02 committed:
Level-8 completed

homuntan02 committed:
A-MoreOOP completed

homuntan02 committed:
A-Packages completed

homuntan02 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

homuntan02 committed:
A-JUnit Completed, added 3 tests.

homuntan02 committed:
Added JavaDocs to classes and methods

homuntan02 committed:
Fixed code to follow coding standard.

homuntan02 committed:
Implemented Level-9 Find

clevon-w commented on src/main/java/duke/parsers/Parser.java:
```diff
@@ -0,0 +1,80 @@
+            return new EndCommand();
+        }
+
+        switch(splitStr[0]) {
+        case "list":
+            return new ListCommand();
+
+        case "mark":
+            if (requestContent.trim() == "") {
+                throw new MissingArgumentException("The index cannot be empty.");
+            }
+            int markIndex = Integer.parseInt(splitStr[1]) - 1;
+            return new MarkCommand(markIndex);
+
+        case "unmark":
+            if (requestContent.trim() == "") {
+                throw new MissingArgumentException("The index cannot be empty.");
+            }
+            int unmarkIndex = Integer.parseInt(splitStr[1]) - 1;
+            return new UnmarkCommand(unmarkIndex);
+
+        case "todo":
+            return new ToDoCommand(requestContent);
+
+        case "deadline":
+            return new DeadlineCommand(requestContent);
+
+        case "event":
+            return new EventCommand(requestContent);
+
+        case "delete":
+            if (requestContent.trim() == "") {
+                throw new MissingArgumentException("The index cannot be empty.");
+            }
+            int index = Integer.parseInt(requestContent.trim()) - 1;
```
Does this handle exceptions thrown if the user passes in something like "delete something"? I think if "something" is passed into `Integer.parseInt` an error will be thrown.

benjamin-wee replied:
I like how you used a switch block as it makes you code much more readable and understandable! The form of the switch block also follows the coding standard!

clevon-w commented on src/main/java/duke/commands/DeadlineCommand.java:
```diff
@@ -0,0 +1,23 @@
+
+    /**
+     * Adds deadline to a task list.
+     */
```
Does this require a JavaDoc? This is the only `execute` method that has a JavaDoc out of all the `Command` subclasses.

clevon-w commented on src/test/java/command/DeadlineCommandTest.java:
```diff
@@ -0,0 +1,34 @@
+        Task actual = this.tasks.showTask(0);
+        Deadline expected = new Deadline("Demo","2002-02-12 19:00");
+        assertEquals(expected, actual);
+    }
```
Is it a good idea to test for errors thrown? So that you can check if the right error is thrown for the right edge cases.

clevon-w reviewed (commented):
Really great work Homun! Keep it up!
I just love the way you structured your repository. It made it so easy to read and understand. I have so much to learn from you!

szejiancheng commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,54 @@
+
+
+/**
+ * The main duke program.
```
Perhaps the header javadoc could be more descriptive!

szejiancheng commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    public static void main(String[] args) {
+        new Duke("./data/").run();
```
Is there a way to make this less system specific? What happens when the filepath isnt valid?

szejiancheng commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,51 @@
+package duke;
+import duke.dukeexceptions.DukeExceptions;
```
Consider separating these with new lines for readability

szejiancheng commented on src/main/java/duke/commands/Command.java:
```diff
@@ -0,0 +1,40 @@
+     *
+     * @param tag to know what command is used.
+     */
+    public Command(String tag) {
```
As the tag corresponds to the subtype of Command, is there something more useful you could store in your tag?

szejiancheng commented on src/main/java/duke/commands/EventCommand.java:
```diff
@@ -0,0 +1,20 @@
+ * Command to add an event to a task list.
+ */
+public class EventCommand extends Command {
+    private String requestContent;
```
Seems like you prefer a line space between variables and methods, consider putting one here?

szejiancheng commented on src/main/java/duke/parsers/Parser.java:
```diff
@@ -0,0 +1,80 @@
+                : "";
+
+        //END
+        if (splitStr[0].equals("bye")) {
```
perhaps consider putting this as a "bye" case in your switch block?

szejiancheng commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,88 @@
+
+
+        } catch (IOException exception) {
+            System.out.println(exception.getMessage());
```
So what happens when an exception is encountered during the try block? Where is it thrown to, and can the program be recovered?

szejiancheng commented on src/main/java/duke/tasklist/TaskList.java:
```diff
@@ -0,0 +1,220 @@
+            int index = i + 1;
+            result += "  " + index + ". " + list[i].toString();
+        }
+        System.out.print(result);
```
Is there a way to allow the Ui class to handle all user input and output instead of printing from this class?

szejiancheng commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,110 @@
+     *
+     * @return the task that has been decoded
+     */
+    public static Task decode(String taskString) throws DukeExceptions {
```
I've noticed that you have a parser class, how would you make it so that it handles more parsing rather than the Task class handling parsing?

szejiancheng reviewed (commented):
Cool structure of Duke! I would suggest that the methods be tweaked such that each class is wholly responsible for one aspect of Duke's functionality for debugging and encapsulation purposes.

benjamin-wee commented on src/main/java/duke/commands/Command.java:
```diff
@@ -0,0 +1,40 @@
+package duke.commands;
```
I like how you used sub packages to maintain your code!

benjamin-wee commented on src/main/java/duke/tasklist/TaskList.java:
```diff
@@ -0,0 +1,220 @@
+        }
+
+        this.len--;
+        String reply = "  Noted. I've removed this task:\n"
+                + "    " + deletedTask.toString()
+                + "  Now you have " + this.len + " tasks in the list.\n";
+        System.out.print(reply);
```
Is it possible to allow the Ui class to handle user input and output here?

benjamin-wee commented on src/main/java/duke/tasks/Deadline.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    /**
+     * Returns the deadline in a formatted string to be saved locally.
+     *
+     * @return the deadline in string format
+     */
+    public String saveTask() {
```
I like how the javadoc is descriptive and helps readers understand the code better!

benjamin-wee reviewed (commented):
Good job in general making your code understandable and highly maintainable! Keep up the good work!

homuntan02 committed:
A-JavaDoc completed

homuntan02 committed:
Merge branch 'branch-A-JavaDoc'

homuntan02 committed:
A-CodingStandard done

homuntan02 committed:
Merge branch 'branch-A-CodingStandard'

homuntan02 committed:
Level-9 completed

homuntan02 committed:
Merge branch 'branch-Level-9'

homuntan02 committed:
Added GUI

homuntan02 committed:
Add Assertions

homuntan02 committed:
Merge pull request #2 from homuntan02/branch-A-Assertions

Add Assertions

homuntan02 committed:
Fixed indentation & added header comments

homuntan02 committed:
Merge pull request #3 from homuntan02/branch-A-CodeQuality

Fixed indentation & added header comments

homuntan02 committed:
Storage: stores and reads data in default file location

User cannot choose where to store their saved data

Including the change_file_location command allows users to choose another
location where they can save and load files without affecting the current
data file.

Users can store the data in new file location whilst keeping a copy of
any other task list in other files.

refer to user guide on how to use this command.

homuntan02 committed:
GUI images used were basic

User might find it boring

Added new images for the GUI

User can find it more fun to use the application

homuntan02 committed:
Updated Ui to include borders

homuntan02 committed:
Update README.md

homuntan02 committed:
Added User Guide and Product Screenshot

homuntan02 committed:
Added project screenshot

homuntan02 committed:
Fixed reply format for GUI

homuntan02 committed:
Fixed Style Check

homuntan02 committed:
Update Ui Screenshot
